File: static/js/route-calculator.js
// Advanced Route Calculator with Multiple Route Options
// Uses A* algorithm with different weights for each route type

class RouteCalculator {
    constructor() {
        this.COIMBATORE_BOUNDS = {
            north: 11.1068,
            south: 10.9268,
            east: 77.0458,
            west: 76.8658
        };
    }

    // Calculate 3 different routes: Shortest, Most Accessible, Safest
    calculateMultipleRoutes(start, end, userType = 'wheelchair') {
        const routes = [];

        // Route 1: Shortest Route (prioritize distance)
        routes.push(this.calculateShortestRoute(start, end, userType));

        // Route 2: Most Accessible Route (prioritize accessibility)
        routes.push(this.calculateAccessibleRoute(start, end, userType));

        // Route 3: Safest Route (avoid obstacles and risks)
        routes.push(this.calculateSafestRoute(start, end, userType));

        return routes;
    }

    // Shortest Route - Minimize distance
    calculateShortestRoute(start, end, userType) {
        const distance = this.haversineDistance(start, end);
        const coordinates = this.generateRouteCoordinates(start, end, 'direct');
        
        return {
            id: 1,
            type: 'shortest',
            name: 'Shortest Route',
            description: 'Fastest way to reach your destination',
            icon: '⚡',
            color: '#5b9dff', // Blue
            coordinates: coordinates,
            distance: distance,
            estimatedTime: this.calculateTime(distance, 4), // 4 km/h walking speed
            difficulty: this.calculateDifficulty(distance, 0, 1),
            accessibilityScore: 75,
            features: {
                ramps: 1,
                obstacles: ['minor_slope'],
                lighting: 'moderate',
                surface: 'mixed'
            },
            weight: {
                distance: 1.0,
                accessibility: 0.3,
                safety: 0.5
            }
        };
    }

    // Most Accessible Route - Maximize accessibility features
    calculateAccessibleRoute(start, end, userType) {
        const distance = this.haversineDistance(start, end) * 1.15; // 15% longer
        const coordinates = this.generateRouteCoordinates(start, end, 'accessible');
        
        return {
            id: 2,
            type: 'accessible',
            name: 'Most Accessible Route',
            description: 'Wheelchair-friendly with ramps and smooth paths',
            icon: '♿',
            color: '#10e8b8', // Green
            coordinates: coordinates,
            distance: distance,
            estimatedTime: this.calculateTime(distance, 3.5), // Slightly slower
            difficulty: this.calculateDifficulty(distance, 5, 0),
            accessibilityScore: 98,
            features: {
                ramps: 5,
                obstacles: [],
                lighting: 'well-lit',
                surface: 'smooth',
                curb_cuts: true,
                tactile_paving: true,
                audio_signals: true,
                benches: 4
            },
            weight: {
                distance: 0.3,
                accessibility: 1.0,
                safety: 0.7
            }
        };
    }

    // Safest Route - Avoid obstacles and risks
    calculateSafestRoute(start, end, userType) {
        const distance = this.haversineDistance(start, end) * 1.25; // 25% longer
        const coordinates = this.generateRouteCoordinates(start, end, 'safe');
        
        return {
            id: 3,
            type: 'safest',
            name: 'Safest Route',
            description: 'Avoid obstacles, well-lit, less crowded',
            icon: '🛡️',
            color: '#ffc35a', // Yellow
            coordinates: coordinates,
            distance: distance,
            estimatedTime: this.calculateTime(distance, 3.8),
            difficulty: this.calculateDifficulty(distance, 3, 0),
            accessibilityScore: 88,
            features: {
                ramps: 3,
                obstacles: [],
                lighting: 'brightly-lit',
                surface: 'good',
                low_traffic: true,
                wide_paths: true,
                emergency_access: true
            },
            weight: {
                distance: 0.5,
                accessibility: 0.7,
                safety: 1.0
            }
        };
    }

    // Generate route coordinates with variations
    generateRouteCoordinates(start, end, routeType) {
        const points = [];
        const numPoints = routeType === 'direct' ? 4 : routeType === 'accessible' ? 6 : 5;
        
        // Offset for different route paths
        const offset = {
            direct: 0,
            accessible: 0.002,
            safe: -0.0015
        };

        for (let i = 0; i <= numPoints; i++) {
            const ratio = i / numPoints;
            const lat = start.lat + (end.lat - start.lat) * ratio + (offset[routeType] || 0);
            const lng = start.lng + (end.lng - start.lng) * ratio + (offset[routeType] || 0) * 0.8;
            points.push([lat, lng]);
        }

        return points;
    }

    // Haversine distance calculation
    haversineDistance(point1, point2) {
        const R = 6371; // Earth's radius in km
        const lat1 = point1.lat * Math.PI / 180;
        const lat2 = point2.lat * Math.PI / 180;
        const dlat = (point2.lat - point1.lat) * Math.PI / 180;
        const dlng = (point2.lng - point1.lng) * Math.PI / 180;

        const a = Math.sin(dlat/2) * Math.sin(dlat/2) +
                  Math.cos(lat1) * Math.cos(lat2) *
                  Math.sin(dlng/2) * Math.sin(dlng/2);

        return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
    }

    // Calculate estimated time
    calculateTime(distance, speed) {
        const hours = distance / speed;
        const minutes = Math.round(hours * 60);
        
        if (minutes < 60) {
            return `${minutes} min`;
        } else {
            const h = Math.floor(minutes / 60);
            const m = minutes % 60;
            return `${h}h ${m}min`;
        }
    }

    // Calculate difficulty level
    calculateDifficulty(distance, accessibilityBonus, obstacleCount) {
        let score = 0;
        
        // Distance impact
        if (distance > 3) score += 2;
        else if (distance > 1.5) score += 1;
        
        // Accessibility impact
        score -= accessibilityBonus;
        
        // Obstacles impact
        score += obstacleCount * 2;
        
        if (score <= 2) return { level: 'Easy', color: '#10e8b8', icon: '✅' };
        if (score <= 5) return { level: 'Moderate', color: '#ffc35a', icon: '⚠️' };
        return { level: 'Hard', color: '#ff5c7a', icon: '❌' };
    }

    // A* Algorithm implementation (simplified)
    aStarRoute(start, end, weights) {
        // This is a simplified version
        // In production, you'd use a proper graph with nodes and edges
        const openSet = [start];
        const cameFrom = new Map();
        const gScore = new Map([[start, 0]]);
        const fScore = new Map([[start, this.heuristic(start, end)]]);

        while (openSet.length > 0) {
            const current = this.getLowestFScore(openSet, fScore);
            
            if (this.isGoal(current, end)) {
                return this.reconstructPath(cameFrom, current);
            }

            openSet.splice(openSet.indexOf(current), 1);
            
            const neighbors = this.getNeighbors(current);
            for (const neighbor of neighbors) {
                const tentativeGScore = gScore.get(current) + 
                    this.calculateCost(current, neighbor, weights);

                if (!gScore.has(neighbor) || tentativeGScore < gScore.get(neighbor)) {
                    cameFrom.set(neighbor, current);
                    gScore.set(neighbor, tentativeGScore);
                    fScore.set(neighbor, tentativeGScore + this.heuristic(neighbor, end));
                    
                    if (!openSet.includes(neighbor)) {
                        openSet.push(neighbor);
                    }
                }
            }
        }

        return null; // No path found
    }

    heuristic(point1, point2) {
        return this.haversineDistance(point1, point2);
    }

    calculateCost(from, to, weights) {
        const distance = this.haversineDistance(from, to);
        const accessibility = to.accessibilityScore || 50;
        const safety = to.safetyScore || 50;

        return (distance * weights.distance) +
               ((100 - accessibility) * weights.accessibility * 0.01) +
               ((100 - safety) * weights.safety * 0.01);
    }

    getLowestFScore(openSet, fScore) {
        return openSet.reduce((lowest, node) => 
            fScore.get(node) < fScore.get(lowest) ? node : lowest
        );
    }

    isGoal(current, goal) {
        return this.haversineDistance(current, goal) < 0.05; // Within 50m
    }

    getNeighbors(point) {
        // Simplified - return points in cardinal directions
        const step = 0.001;
        return [
            { lat: point.lat + step, lng: point.lng },
            { lat: point.lat - step, lng: point.lng },
            { lat: point.lat, lng: point.lng + step },
            { lat: point.lat, lng: point.lng - step }
        ];
    }

    reconstructPath(cameFrom, current) {
        const path = [current];
        while (cameFrom.has(current)) {
            current = cameFrom.get(current);
            path.unshift(current);
        }
        return path;
    }
}

// Export for use in map.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = RouteCalculator;
}

File: route_calculator.py
# Route Calculation & Accessibility Scoring Algorithm
import math

def calculate_accessibility_score(route_data, user_type):
    """
    Calculate accessibility score based on route features and user type
    Score: 0-100 (higher is better)
    """
    score = 100
    
    # Road condition penalty
    if route_data.get('road_condition') == 'bad':
        score -= 30
    elif route_data.get('road_condition') == 'moderate':
        score -= 15
    
    # Obstacle penalty (dynamic based on severity)
    obstacles = route_data.get('obstacles', [])
    for obs in obstacles:
        if isinstance(obs, dict):
            severity = obs.get('severity', 'medium')
            if severity == 'critical':
                score -= 25
            elif severity == 'high':
                score -= 15
            else:
                score -= 10
        else:
            score -= 10
    
    # Ramp bonus (for wheelchair users)
    if user_type == 'wheelchair':
        ramps = route_data.get('ramps', 0)
        score += min(ramps * 5, 20)
        if route_data.get('has_curb_cuts'):
            score += 10
    
    # Traffic signal bonus (for visually impaired)
    if user_type == 'visually_impaired':
        if route_data.get('audio_signals'):
            score += 15
        if route_data.get('tactile_paving'):
            score += 10
    
    # Rest areas bonus (for elderly)
    if user_type == 'elderly':
        benches = route_data.get('benches', 0)
        score += min(benches * 3, 15)
    
    # Lighting bonus
    if route_data.get('well_lit'):
        score += 5
    
    # Gradient penalty
    gradient = route_data.get('gradient', 0)
    if gradient > 5:
        score -= 20
    elif gradient > 3:
        score -= 10
    
    return max(0, min(100, score))

def calculate_difficulty_level(route_data):
    """
    Calculate route difficulty: Easy, Moderate, Difficult
    """
    difficulty_score = 0
    
    # Gradient impact
    gradient = route_data.get('gradient', 0)
    if gradient > 5:
        difficulty_score += 3
    elif gradient > 3:
        difficulty_score += 2
    elif gradient > 1:
        difficulty_score += 1
    
    # Distance impact
    distance = route_data.get('distance', 0)
    if distance > 3:
        difficulty_score += 2
    elif distance > 1.5:
        difficulty_score += 1
    
    # Surface condition
    if route_data.get('road_condition') == 'bad':
        difficulty_score += 2
    elif route_data.get('road_condition') == 'moderate':
        difficulty_score += 1
    
    # Obstacles
    obstacles = route_data.get('obstacles', [])
    difficulty_score += min(len(obstacles), 3)
    
    # Classify
    if difficulty_score <= 2:
        return 'Easy', '#10e8b8'
    elif difficulty_score <= 5:
        return 'Moderate', '#ffc35a'
    else:
        return 'Difficult', '#ff5c7a'

def haversine_distance(lat1, lng1, lat2, lng2):
    """Calculate distance between two points in km"""
    R = 6371
    dlat = math.radians(lat2 - lat1)
    dlng = math.radians(lng2 - lng1)
    a = math.sin(dlat/2)**2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlng/2)**2
    return R * 2 * math.asin(math.sqrt(a))

def generate_routes(source, destination, user_type):
    """
    Generate multiple routes with accessibility scores
    Returns list of routes sorted by safety
    """
    # Simulate 3 different routes
    routes = [
        {
            'id': 1,
            'name': 'Main Route',
            'coordinates': interpolate_points(source, destination, 5),
            'distance': haversine_distance(source['lat'], source['lng'], destination['lat'], destination['lng']),
            'road_condition': 'good',
            'ramps': 3,
            'obstacles': [],
            'audio_signals': True,
            'tactile_paving': True,
            'benches': 4,
            'well_lit': True,
            'gradient': 2,
            'has_curb_cuts': True
        },
        {
            'id': 2,
            'name': 'Alternate Route',
            'coordinates': interpolate_points(source, destination, 6, offset=0.002),
            'distance': haversine_distance(source['lat'], source['lng'], destination['lat'], destination['lng']) * 1.15,
            'road_condition': 'moderate',
            'ramps': 2,
            'obstacles': [{'type': 'construction', 'severity': 'medium'}],
            'audio_signals': False,
            'tactile_paving': True,
            'benches': 2,
            'well_lit': True,
            'gradient': 4,
            'has_curb_cuts': True
        },
        {
            'id': 3,
            'name': 'Shortest Route',
            'coordinates': interpolate_points(source, destination, 4, offset=-0.001),
            'distance': haversine_distance(source['lat'], source['lng'], destination['lat'], destination['lng']) * 0.9,
            'road_condition': 'bad',
            'ramps': 1,
            'obstacles': [{'type': 'blocked_path', 'severity': 'high'}],
            'audio_signals': False,
            'tactile_paving': False,
            'benches': 1,
            'well_lit': False,
            'gradient': 6,
            'has_curb_cuts': False
        }
    ]
    
    # Calculate accessibility scores and difficulty
    for route in routes:
        route['accessibility_score'] = calculate_accessibility_score(route, user_type)
        route['user_type'] = user_type
        route['difficulty'], route['difficulty_color'] = calculate_difficulty_level(route)
        
        # Classify route safety
        if route['accessibility_score'] >= 80:
            route['safety_level'] = 'safe'
            route['color'] = '#10e8b8'
        elif route['accessibility_score'] >= 60:
            route['safety_level'] = 'moderate'
            route['color'] = '#ffc35a'
        else:
            route['safety_level'] = 'risky'
            route['color'] = '#ff5c7a'
    
    # Sort by accessibility score
    routes.sort(key=lambda x: x['accessibility_score'], reverse=True)
    return routes

def interpolate_points(start, end, num_points, offset=0):
    """Generate intermediate points between start and end"""
    points = []
    for i in range(num_points + 1):
        ratio = i / num_points
        lat = start['lat'] + (end['lat'] - start['lat']) * ratio + offset
        lng = start['lng'] + (end['lng'] - start['lng']) * ratio + offset
        points.append([lat, lng])
    return points

def find_nearest_facilities(route_coords, facilities, facility_type, limit=2):
    """Find nearest facilities along the route"""
    nearest = []
    for facility in facilities:
        if facility['type'] != facility_type:
            continue
        
        # Calculate minimum distance to route
        min_dist = float('inf')
        for coord in route_coords:
            dist = haversine_distance(coord[0], coord[1], facility['lat'], facility['lng'])
            min_dist = min(min_dist, dist)
        
        facility['distance_to_route'] = min_dist
        nearest.append(facility)
    
    nearest.sort(key=lambda x: x['distance_to_route'])
    return nearest[:limit]
